compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp

```diff
@@ -32,58 +33,67 @@ void AMDAIEFuseFillIntoForallPass::runOnOperation() {
+      extractSliceOp = nxt;
     }
-    return WalkResult::advance();
-  });
+  }
 
-  if (!tileableProducer) {
-    LLVM_DEBUG(llvm::dbgs() << "There is no producer op to be fused.\n");
+  if (extractSliceOp) {
+    LoopLikeOpInterface loops =
+        cast<LoopLikeOpInterface>(forallOp.getOperation());
+
+    // Materialize the slice of the producer in place.
+    std::optional<scf::SCFFuseProducerOfSliceResult> fusedProducer =
+        scf::tileAndFuseProducerOfSlice(rewriter, extractSliceOp,
+                                        MutableArrayRef(&loops, 1));
+    if (!fusedProducer) {
+      funcOp->emitOpError("Failed to fuse fill op into forall loop.");
+      return signalPassFailure();
+    }
     return;
   }
 
-  // Search the first use by a scf::ForallOp user.
-  scf::ForallOp forallOp;
-  auto itProducerUses =
-      llvm::find_if(tileableProducer->getUses(), [&](OpOperand &use) {
-        forallOp = dyn_cast<scf::ForallOp>(use.getOwner());
-        return forallOp;
+  // In the case where there are no extract_slice ops, we manually create the
+  // fill at the beginning of the forall body.
+  assert(!extractSliceOp);
+  rewriter.setInsertionPointToStart(forallOp.getBody());
+  Value scalar = fillOp.value();
+  Location loc = fillOp.getLoc();
+  auto fusedFill = rewriter.create<linalg::FillOp>(loc, scalar, bbArg);
+  rewriter.replaceUsesWithIf(
+      bbArg, fusedFill.getResult(0), [&](OpOperand &operand) {
+        Operation *owner = operand.getOwner();
+        if (owner == fusedFill) {
+          return false;
+        } else if (isa<tensor::ParallelInsertSliceOp>(owner)) {
+          return false;
+        } else {
+          return true;
+        }
       });
-  if (!forallOp) {
-    LLVM_DEBUG(llvm::dbgs() << "There is no forall Op.\n");
-    return;
-  }
 
-  // Search the producer slices accessed within the Forall op.
-  OpOperand *pUse = &(*itProducerUses);
-  BlockArgument bbArg = forallOp.getTiedBlockArgument(pUse);
-
-  auto itBBArgUsers = llvm::find_if(bbArg.getUsers(), [&](Operation *user) {
-    auto sliceOp = dyn_cast<tensor::ExtractSliceOp>(user);
-    return sliceOp;
-  });
-  if (itBBArgUsers == bbArg.getUsers().end()) {
-    funcOp->emitOpError("There is no extract tensor slice.");
-    return signalPassFailure();
-  }
-  auto sliceOpToTile = cast<tensor::ExtractSliceOp>(*itBBArgUsers);
-
-  LoopLikeOpInterface loops =
-      cast<LoopLikeOpInterface>(forallOp.getOperation());
-
-  // Materialize the slice of the producer in place.
-  std::optional<scf::SCFFuseProducerOfSliceResult> fusedProducer =
-      scf::tileAndFuseProducerOfSlice(rewriter, sliceOpToTile,
-                                      MutableArrayRef(&loops, 1));
-  if (!fusedProducer) {
-    funcOp->emitOpError("Failed to fuse fill op into forall loop.");
-    return signalPassFailure();
-  }
+  // Do not use the result of the old fill.
+  rewriter.replaceAllUsesWith(fillOp.getResults()[0], fillOp.getOutputs()[0]);
```

Comment:
Can't this be included in the above replaceUsesWithIf?

Reply:
No, different ops